### PR TITLE
PDO auto instrumentation #1

### PR DESCRIPTION
**Pull request by pdelewski**
Initial commit of PDO auto instrumentation

### CODE REVIEW

**Comment by pdelewski** [comment]
> don't forget to add this to `.gitsplit.yml`. I can create (or show you how) a read-only repo for it to be pushed to. Then, we can create an entry in packagist for this.

ok, thank you

File: src/Instrumentation/PDO/composer.json [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  "keywords": ["opentelemetry", "otel", "open-telemetry", "tracing", "pdo", "instrumentation"],
+  "type": "library",
+  "homepage": "https://opentelemetry.io/docs/php",
+  "readme": "./README.md",
+  "license": "Apache-2.0",
+  "minimum-stability": "dev",
+  "require": {
+    "php": "^8.0",
+    "ext-otel_instrumentation": "*",
+    "open-telemetry/api": "*",
```
Comment (brettmc) [inline src/Instrumentation/PDO/composer.json:13]:
```suggestion
    "open-telemetry/api": "^1",
```

File: src/Instrumentation/PDO/src/PDOInstrumentation.php [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+            \PDO::class,
+            '__construct',
+            pre: static function (\PDO $pdo, array $params, string $class, string $function, ?string $filename, ?int $lineno) use ($instrumentation) {
+                /** @psalm-suppress ArgumentTypeCoercion */
+                $builder = self::makeBuilder($instrumentation, 'PDO::__construct', $function, $class, $filename, $lineno)
+                    ->setSpanKind(SpanKind::KIND_CLIENT)
+                    ->setAttribute(TraceAttributes::DB_CONNECTION_STRING, $params[0] ?? 'unknown')
+                    ->setAttribute(TraceAttributes::DB_USER, $params[1] ?? 'unknown')
+                    // parse connection string
+                    ->setAttribute(TraceAttributes::DB_SYSTEM, 'mysql');
```
Comment (brettmc) [inline src/Instrumentation/PDO/src/PDOInstrumentation.php:35]: can this be implemented to get real value from connection string?

File: src/Instrumentation/PDO/src/PDOInstrumentation.php [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+                self::end($exception);
+            }
+        );
+
+        hook(
+            \PDOStatement::class,
+            'fetchAll',
+            pre: static function (\PDOStatement $statement, array $params, string $class, string $function, ?string $filename, ?int $lineno) use ($instrumentation) {
+                /** @psalm-suppress ArgumentTypeCoercion */
+                $builder = self::makeBuilder($instrumentation, 'PDOStatement::fetch', $function, $class, $filename, $lineno)
```
Comment (brettmc) [inline src/Instrumentation/PDO/src/PDOInstrumentation.php:97]:
```suggestion
                $builder = self::makeBuilder($instrumentation, 'PDOStatement::fetchAll', $function, $class, $filename, $lineno)
```

File: src/Instrumentation/PDO/src/PDOInstrumentation.php [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+                self::end($exception);
+            }
+        );
+
+        hook(
+            \PDOStatement::class,
+            'execute',
+            pre: static function (\PDOStatement $statement, array $params, string $class, string $function, ?string $filename, ?int $lineno) use ($instrumentation) {
+                /** @psalm-suppress ArgumentTypeCoercion */
+                $builder = self::makeBuilder($instrumentation, 'PDOStatement::fetch', $function, $class, $filename, $lineno)
```
Comment (brettmc) [inline src/Instrumentation/PDO/src/PDOInstrumentation.php:113]:
```suggestion
                $builder = self::makeBuilder($instrumentation, 'PDOStatement::execute', $function, $class, $filename, $lineno)
```

File: src/Instrumentation/PDO/src/PDOInstrumentation.php (hunk 2 above)
Comment (pdelewski) [reply src/Instrumentation/PDO/src/PDOInstrumentation.php:35]: yes, right. I forgot about this gap.

File: src/Instrumentation/PDO/composer.json [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  "type": "library",
+  "homepage": "https://opentelemetry.io/docs/php",
+  "readme": "./README.md",
+  "license": "Apache-2.0",
+  "minimum-stability": "dev",
+  "require": {
+    "php": "^8.0",
+    "ext-otel_instrumentation": "*",
+    "open-telemetry/api": "^1",
+    "nyholm/psr7": "^1",
```
Comment (brettmc) [inline src/Instrumentation/PDO/composer.json:14]: I think nyholm/* is not required for this instrumentation

**Review by brettmc — APPROVED** [review]
Approved, merge when you're ready. I've created https://github.com/opentelemetry-php/contrib-auto-pdo so the git-split has a destination...

File: src/Instrumentation/PDO/composer.json (hunk 5 above)
Comment (pdelewski) [reply src/Instrumentation/PDO/composer.json:14]: Yes, right. Will remove it.

**Comment by pdelewski** [comment]
> Approved, merge when you're ready. I've created https://github.com/opentelemetry-php/contrib-auto-pdo so the git-split has a destination...

ok, thank you

**Comment by pdelewski** [comment]
@brettmc For dsn parsing I added additional dependency https://github.com/Nyholm/dsn. If you have something better or you think that we should implement this by ourselves just let me know (will do that in following PR, however for me it would be reinventing the wheel).

**Comment by pdelewski** [comment]
> @brettmc For dsn parsing I added additional dependency https://github.com/Nyholm/dsn. If you have something better or you think that we should implement this by ourselves just let me know (will do that in following PR, however for me it would be reinventing the wheel).

Hmm, I see that sdk contains dsn parser

File: src/Instrumentation/PDO/src/PDOInstrumentation.php [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        hook(
+            \PDO::class,
+            '__construct',
+            pre: static function (\PDO $pdo, array $params, string $class, string $function, ?string $filename, ?int $lineno) use ($instrumentation) {
+                /** @psalm-suppress ArgumentTypeCoercion */
+                $builder = self::makeBuilder($instrumentation, 'PDO::__construct', $function, $class, $filename, $lineno)
+                    ->setSpanKind(SpanKind::KIND_CLIENT)
+                    ->setAttribute(TraceAttributes::DB_CONNECTION_STRING, $params[0] ?? 'unknown')
+                    ->setAttribute(TraceAttributes::DB_USER, $params[1] ?? 'unknown')
+                    ->setAttribute(TraceAttributes::DB_SYSTEM, self::getDB($params[0]));
```
Comment (Nevay) [inline src/Instrumentation/PDO/src/PDOInstrumentation.php:35]: You can use `$pdo->getAttribute(PDO::ATTR_DRIVER_NAME)` to set `db.system` in the post hook.

File: src/Instrumentation/PDO/src/PDOInstrumentation.php [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+        $instrumentation = new CachedInstrumentation('io.opentelemetry.contrib.php.pdo');
+
+        hook(
+            \PDO::class,
+            '__construct',
+            pre: static function (\PDO $pdo, array $params, string $class, string $function, ?string $filename, ?int $lineno) use ($instrumentation) {
+                /** @psalm-suppress ArgumentTypeCoercion */
+                $builder = self::makeBuilder($instrumentation, 'PDO::__construct', $function, $class, $filename, $lineno)
+                    ->setSpanKind(SpanKind::KIND_CLIENT)
+                    ->setAttribute(TraceAttributes::DB_CONNECTION_STRING, $params[0] ?? 'unknown')
```
Comment (Nevay) [inline src/Instrumentation/PDO/src/PDOInstrumentation.php:33]: Should check that `$class === PDO::class`, subclasses might declare different parameters.

File: src/Instrumentation/PDO/src/PDOInstrumentation.php [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+                $span = $builder->startSpan();
+                Context::storage()->attach($span->storeInContext($parent));
+            },
+            post: static function (\PDO $pdo, array $params, mixed $statement, ?Throwable $exception) {
+                self::end($exception);            }
+        );
+
+        hook(
+            \PDOStatement::class,
+            'fetch',
```
Comment (Nevay) [inline src/Instrumentation/PDO/src/PDOInstrumentation.php:78]: IMO we shouldn't create a span per `::fetch()` etc.; large result sets will lead to a lot of spans.

Comment (pdelewski) [reply src/Instrumentation/PDO/src/PDOInstrumentation.php:78]: I was thinking about and plan to change this behaviour in subsequent PR(s).

File: src/Instrumentation/PDO/src/PDOInstrumentation.php (hunk 6 above)
Comment (pdelewski) [reply src/Instrumentation/PDO/src/PDOInstrumentation.php:35]: good point

**Pull request merged**